core/transaction-pool/graph/src/pool.rs

```diff
@@ -152,8 +152,51 @@ impl<B: ChainApi> Pool<B> {
 		Ok(watcher)
 	}
 
+	/// Prunes ready transactions.
+	///
+	/// Each extrinsic from the list is removed from the pool by pruning a tag that it provides.
+	/// If it's known we take the provided tags we already have.
+	/// If it's not in the pool we query the runtime to get info about the tags.
+	/// Invalid extrinsics (most likely inherent) are silently ignored.
+	pub fn prune(&self, at: &BlockId<B::Block>, parent: &BlockId<B::Block>, extrinsics: &[ExtrinsicFor<B>]) -> Result<(), B::Error> {
+		let mut tags = Vec::with_capacity(extrinsics.len());
+		// re use tags of known transactions to avoid calling into runtime
+		let ready = self.pool.read().by_hash(extrinsics.iter().map(|xt| self.api.hash(xt)));
+		let all = extrinsics.iter().zip(ready.clone());
+
+		for (xt, existing) in all {
```

Comment:
I would be more than happy to see step-by-step comments describing the implementation. Unfortunately, it's really hard to follow the logic here. Also, I'd consider renaming `tx` and `xt` to something more verbose.